x-pack/plugins/lens/public/trigger_actions/open_lens_config/in_app_embeddable_edit/in_app_embeddable_edit_action_helpers.tsx

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import React from 'react';
+import ReactDOM from 'react-dom';
+import type { CoreStart } from '@kbn/core/public';
+import { isOfAggregateQueryType } from '@kbn/es-query';
+import { toMountPoint } from '@kbn/kibana-react-plugin/public';
+import { IncompatibleActionError } from '@kbn/ui-actions-plugin/public';
+import type { LensPluginStartDependencies } from '../../../plugin';
+import type { TypedLensByValueInput } from '../../../embeddable/embeddable_component';
+import { extractReferencesFromState } from '../../../utils';
+import type { LensChartLoadEvent } from './types';
+
+export function isEmbeddableEditActionCompatible(
+  core: CoreStart,
+  attributes: TypedLensByValueInput['attributes']
+) {
+  // for ES|QL is compatible only when advanced setting is enabled
+  const query = attributes.state.query;
+  return isOfAggregateQueryType(query) ? core.uiSettings.get('discover:enableESQL') : true;
+}
+
+export async function executeEditEmbeddableAction({
+  deps,
+  core,
+  attributes,
+  lensEvent,
+  container,
+  onUpdate,
+  onApply,
+  onCancel,
+}: {
+  deps: LensPluginStartDependencies;
+  core: CoreStart;
+  attributes: TypedLensByValueInput['attributes'];
+  lensEvent: LensChartLoadEvent;
+  container?: HTMLElement | null;
+  onUpdate: (newAttributes: TypedLensByValueInput['attributes']) => void;
+  onApply?: () => void;
+  onCancel?: () => void;
+}) {
+  const isCompatibleAction = isEmbeddableEditActionCompatible(core, attributes);
+  if (!isCompatibleAction) {
+    throw new IncompatibleActionError();
+  }
+
+  const { getEditLensConfiguration, getVisualizationMap, getDatasourceMap } = await import(
+    '../../../async_services'
+  );
+  const visualizationMap = getVisualizationMap();
+  const datasourceMap = getDatasourceMap();
+  const query = attributes.state.query;
+  const activeDatasourceId = isOfAggregateQueryType(query) ? 'textBased' : 'formBased';
```

Comment:
I see your comment above about the shortcomings of `getActiveDatasourceIdFromDoc` but I wonder if it would be too much trouble to update it in this PR? Otherwise, maybe we could knock this off as a quick follow-up! As it stands, users are required to specify a query to the config builder to use ES|QL AFAICT

Reply:
No this activeDatasourceId check here is not due to my comment above. It is required by the inline editing component and getting the query is one way to get the datasourceId. It is irrelevant with what the getActiveDatasourceIdFromDoc does. The ESQL query is super important to exist in the state.